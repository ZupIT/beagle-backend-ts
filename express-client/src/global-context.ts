import { ContextNode } from '@zup-it/beagle-backend-core'

export class GlobalContext<T> extends ContextNode<T> {
  constructor() {
    super('globalContext')
  }
}
