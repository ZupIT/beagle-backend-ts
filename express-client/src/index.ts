export { BeagleApp } from './beagle-app'
export { Screen, ScreenRequest } from './screen'
export { RouteMap } from './route'
