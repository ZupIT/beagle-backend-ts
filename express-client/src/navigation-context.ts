import { ContextNode } from '@zup-it/beagle-backend-core'

export class NavigationContext<T> extends ContextNode<T> {
  constructor() {
    super('navigationContext')
  }
}
