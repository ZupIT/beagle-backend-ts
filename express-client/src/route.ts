import { HttpMethod } from '@zup-it/beagle-backend-core'
import { Screen } from './screen'

export interface RouteConfig {
  method?: HttpMethod,
  path: string,
}

export type RouteMap = Record<string, Screen | { method: HttpMethod, screen: Screen }>
