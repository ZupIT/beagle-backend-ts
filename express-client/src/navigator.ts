import { FC } from '@zup-it/beagle-backend-core'
import {
  pushView, pushStack, popToView, popView, popStack, resetApplication, resetStack, Route,
} from '@zup-it/beagle-backend-core/actions'
import { forEach } from 'lodash'
import {
  ControllerId, PopStackAction, PopToViewAction, PopViewAction, PushStackAction,
  PushViewAction, ResetApplicationAction, ResetStackAction
} from './model/navigation'
import { RouteMap, RouteConfig } from './route'
import { ScreenNavigation } from './screen'

const navigationActions = { pushView, pushStack, popToView, popView, resetApplication, resetStack, popStack }

interface GenericRemoteNavigation {
  type: keyof typeof navigationActions,
  screen?: FC<any>,
  properties?: ScreenNavigation<any, any> & ControllerId,
}

/**
 * See the property `navigation` in the interface `ScreenProps` for a detailed description of the Navigator.
 *
 * The instance of the current navigator is injected by the BeagleApplication for every registered Screen. See the
 * example below:
 *
 * ```
 * const MyScreen: Screen<Type> = ({ navigator }) => {
 *   // ...
 * }
 * ```
 */
export class Navigator {
  /**
   * @param routeMap the same routeMap received by BeagleApplication.
   * @param basePath the same basePath received by BeagleApplication options.
   */
  constructor(routeMap: RouteMap, private basePath = '') {
    this.screenMap = new Map()
    forEach(routeMap, (value, key) => this.screenMap.set(
      typeof value === 'function' ? value : value.screen,
      { path: key, method: typeof value === 'function' ? undefined : value.method },
    ))
  }

  private screenMap: Map<FC<any>, RouteConfig>

  private getPathAndMethod(screen: FC<any>) {
    if (!this.screenMap.has(screen)) {
      throw new Error(`Couldn't find any route corresponding to "${screen}". Are you sure you registered it via BeagleApp.addScreen?`)
    }
    return this.screenMap.get(screen)!
  }

  private buildUrl(path: string, routeParams: Record<string, string> = {}) {
    return path.replace(/:(\w+)/g, (_, name) => name in routeParams ? routeParams[name] : `:${name}`)
  }

  private buildRoute({ type, screen, properties = {} }: GenericRemoteNavigation) {
    if (type === 'popView' || type === 'popStack') return undefined

    const { routeParams, headers, body, shouldPrefetch, fallback } = properties
    const { path, method } = this.getPathAndMethod(screen!)
    const url = this.buildUrl(`${this.basePath}${path}`, routeParams ?? {})
    if (type === 'popToView') return url

    const httpAdditionalData = method || headers || body ? { method, headers, body } : undefined

    return {
      url,
      httpAdditionalData,
      shouldPrefetch,
      fallback,
    } as Route
  }

  private navigateRemote(navigation: GenericRemoteNavigation) {
    return navigationActions[navigation.type]({
      // @ts-ignore todo: check this later
      route: this.buildRoute(navigation),
      // @ts-ignore the type of the analytics is not right considering we can use the operator ".", e.g. "route.url"
      analytics: navigation.properties?.analytics,
      navigationContext: navigation.properties?.navigationContext,
      controllerId: navigation.properties?.controllerId,
    })
  }

  /**
   * Adds a new stack to the navigator with the provided route.
   *
   * @param screen the screen (functional component) to navigate to.
   * @param properties the data to send with this navigation (and analytics).
   * @returns an instance of Action.
   */
  pushStack: PushStackAction = (screen, properties) => this.navigateRemote({ type: 'pushStack', screen, properties })

  /**
   * Pops the current stack, going back to the last route of the previous stack.
   *
   * @param properties the navigation context to set and analytics.
   * @returns an instance of Action.
   */
  popStack: PopStackAction = (properties) => this.navigateRemote({ type: 'popStack', properties })

  /**
   * Adds the provided route to the current navigation stack.
   *
   * @param screen the screen (functional component) to navigate to.
   * @param properties the data to send with this navigation (and analytics).
   * @returns an instance of Action.
   */
  pushView: PushViewAction = (screen, properties) => this.navigateRemote({ type: 'pushView', screen, properties })

  /**
   * Goes back to the previous route.
   *
   * @param properties the navigation context to set and analytics.
   * @returns an instance of Action.
   */
  popView: PopViewAction = (properties) => this.navigateRemote({ type: 'popView', properties })

  /**
   * Goes back to the route identified by the string passed as parameter. If the route doesn't exist in the current
   * navigation stack, nothing happens.
   *
   * @param screen the screen (functional component) to go back to.
   * @param properties the data to send with this navigation (and analytics).
   * @returns an instance of Action.
   */
  popToView: PopToViewAction = (screen, properties) => this.navigateRemote({ type: 'popToView', screen, properties })

  /**
   * Removes the current navigation stack and adds a new one with the provided route.
   *
   * @param screen the screen (functional component) to navigate to.
   * @param properties the data to send with this navigation (and analytics).
   * @returns an instance of Action.
   */
  resetStack: ResetStackAction = (screen, properties) => this.navigateRemote({ type: 'resetStack', screen, properties })

  /**
   * Removes all the navigation stacks and adds a new one with the provided route.
   *
   * @param screen the screen (functional component) to navigate to.
   * @param properties the data to send with this navigation (and analytics).
   * @returns an instance of Action.
   */
  resetApplication: ResetApplicationAction = (screen, properties) => (
    this.navigateRemote({ type: 'resetApplication', screen, properties })
  )
}
