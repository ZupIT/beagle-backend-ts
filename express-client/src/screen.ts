import { Request, Response } from 'express'
import { WithAnalytics, AnyContextNode, Component } from '@zup-it/beagle-backend-core'
import { Navigator } from './navigator'

export interface RequestWithCustomHeaders<RouteParams = any, Headers = any, Body = any, Query = any>
  extends Request<RouteParams, any, Body, Query> {
  headers: Request['headers'] & Headers,
}

export interface ScreenRequest {
  headers: Record<string, string>,
  routeParams: Record<string, string>,
  query: Record<string, string>,
  body: any,
  navigationContext: any,
}

interface ScreenProps<T extends ScreenRequest> {
  navigationContext: AnyContextNode<T['navigationContext']>,
  request: RequestWithCustomHeaders<T['routeParams'], T['headers'], T['body'], T['query']>,
  response: Response,
  navigator: Navigator,
}

export type Screen<T extends ScreenRequest = any> = (props: ScreenProps<T>) => JSX.Element

export interface ScreenNavigation<T extends ScreenRequest> extends WithAnalytics {
  routeParams?: T['routeParams'],
  headers?: T['headers'],
  body?: T['body'],
  query?: T['query'],
  navigationContext?: T['navigationContext'],
  shouldPrefetch?: boolean,
  fallback?: Component,
}
