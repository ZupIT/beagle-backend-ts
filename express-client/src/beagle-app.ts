import { forEach } from 'lodash'
import { Express } from 'express'
import { HttpMethod, serialize, FC, ContextNode, AnyContextNode } from '@zup-it/beagle-backend-core'
import { Screen } from './screen'
import { Navigator } from './navigator'

interface Options {
  responseHeaders?: Record<string, string>,
  basePath?: string,
}

interface ScreenProperties {
  method?: HttpMethod,
  path: string,
  screen: FC<Screen>,
}

export class BeagleApp<GlobalContextType = any> {
  constructor(private express: Express, options: Options = {}) {
    this.responseHeaders = options.responseHeaders ?? {}
    this.basePath = options.basePath ?? ''
    this.navigator = new Navigator(this.basePath)
  }

  private responseHeaders: Record<string, any>
  private basePath: string
  // The casting below shouldn't be necessary. I don't know why TS complains.
  readonly globalContext = new ContextNode('globalContext') as unknown as AnyContextNode<GlobalContextType>
  readonly navigator: Navigator

  addScreen(properties: ScreenProperties): void {
    const { method = 'get', path, screen } = properties
    this.navigator.register(properties.screen, properties)
    this.express[method](`${this.basePath}${path}`, (req, res) => {
      res.type('application/json')
      forEach(this.responseHeaders, (key, value) => res.setHeader(key, value))
      const componentTree = screen({
        request: req,
        response: res,
        navigationContext: new ContextNode('navigationContext'),
      })
      res.send(serialize(componentTree))
    })
  }
}
