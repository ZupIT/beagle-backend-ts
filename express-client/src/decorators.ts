import { HttpMethod } from '@zup-it/beagle-backend-core'
import { BeagleApp } from './beagle-app'
import { ScreenClass } from './screen'

let app: BeagleApp | undefined
let screenQueue: Parameters<BeagleApp['addScreen']>[] = []

export const Beagle = () => (AppFactory: { new(): BeagleApp }) => {
  app = new AppFactory()
  if (screenQueue.length) {
    screenQueue.forEach(item => app!.addScreen(...item))
    screenQueue = []
  }
}

function register(method: HttpMethod, path: string, ScreenFactory: ScreenClass) {
  if (!app) screenQueue.push([method, path, ScreenFactory])
  else app.addScreen(method, path, ScreenFactory)
}

export const GetScreen = (path: string) => (ScreenFactory: ScreenClass) => register('get', path, ScreenFactory)
export const PostScreen = (path: string) => (ScreenFactory: ScreenClass) => register('post', path, ScreenFactory)
export const PutScreen = (path: string) => (ScreenFactory: ScreenClass) => register('put', path, ScreenFactory)
export const PatchScreen = (path: string) => (ScreenFactory: ScreenClass) => register('patch', path, ScreenFactory)
export const DeleteScreen = (path: string) => (ScreenFactory: ScreenClass) => register('delete', path, ScreenFactory)
