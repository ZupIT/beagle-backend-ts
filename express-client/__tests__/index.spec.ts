import { mapValues } from 'lodash'
import * as express from 'src'

describe('Express module', () => {
  it('should export', () => expect(mapValues(express, v => typeof v)).toMatchSnapshot())
})
