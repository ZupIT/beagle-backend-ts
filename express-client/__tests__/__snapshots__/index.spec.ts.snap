// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Express module should export 1`] = `
Object {
  "BeagleApp": "function",
  "isMobile": "function",
  "isWeb": "function",
}
`;
